src/java/org/apache/cassandra/locator/AbstractReplicationStrategy.java

```diff
@@ -457,4 +433,48 @@ protected void validateExpectedOptions() throws ConfigurationException
                 throw new ConfigurationException(String.format("Unrecognized strategy option {%s} passed to %s for keyspace %s", key, getClass().getSimpleName(), keyspaceName));
         }
     }
+
+    static class ReplicaCache<K, V>
```

Comment:
It looks like we can slightly improve this by introducing a versioned replicas map (i.e. a class that would hold a final cached version, and final map of replicas), that would be held in an atomic reference. Then, we don't need so synchronise on `this`, and can just create and return a new version of cached replicas.